[10.0][FIX] mail_tracking_mailgun: Avoid key dict errors

chienandalu opened the pull request:
- Avoids error if dictionary keys doesn't exist.

cc @Tecnativa

chienandalu committed:
[10.0][FIX] mail_tracking_mailgun: Avoid key dict errors

pedrobaeza commented on mail_tracking_mailgun/models/mail_tracking_email.py:
```diff
@@ -240,9 +240,9 @@ def action_manual_check_mailgun(self):
             for item in content["items"]:
                 if not self.env['mail.tracking.event'].search(
-                        [('mailgun_id', '=', item["id"])]):
+                        [('mailgun_id', '=', item.get("id"))]):
```
You can't enter here if there's no ID, but I don't know why this can happen

chienandalu replied:
Well, if there's an ok response there should allways be an id.

pedrobaeza replied:
But again, if there's no ID, you shouldn't perform any search IMO.

pedrobaeza commented on mail_tracking_mailgun/models/mail_tracking_email.py:
```diff
@@ -240,9 +240,9 @@ def action_manual_check_mailgun(self):
+                        [('mailgun_id', '=', item.get("id"))]):
                     mapped_event_type = self._mailgun_event_type_mapping.get(
-                        item["event"]) or False
+                        item.get("event", False))
```
Which type gets in this case?

chienandalu replied:
Right, there was an error... :confused:

pedrobaeza commented on mail_tracking_mailgun/models/res_partner.py:
```diff
@@ -30,7 +30,7 @@ def _email_bounced_set(self, reason, event):
                      'Reason: %s\n'
                      'Event: %s') % (partner.email, reason,
-                                     event['Message-Id'] or '')
+                                     event.get('Message-Id', ''))
```
Put `_('unknown')` as fallback

chienandalu replied:
Done

pedrobaeza reviewed (changes requested):
Change module verison

pedrobaeza commented on mail_tracking_mailgun/models/mail_tracking_email.py:
```diff
@@ -240,9 +242,9 @@ def action_manual_check_mailgun(self):
+                        [('mailgun_id', '=', item.get("id"))]):
                     mapped_event_type = self._mailgun_event_type_mapping.get(
-                        item["event"]) or False
+                        item.get("event"), False)
```
If no event, should you launch `event_create` any way?

pedrobaeza commented on mail_tracking_mailgun/models/res_partner.py:
```diff
@@ -30,7 +30,7 @@ def _email_bounced_set(self, reason, event):
                      'Reason: %s\n'
                      'Event: %s') % (partner.email, reason,
-                                     event['Message-Id'] or '')
+                                     event.get('Message-Id', _('unknown')))
```
Include translation for this term

yajo committed:
[FIX] mail_tracking_mailgun: Avoid errors without event

When this method is called without an event (a.k.a. `event=None`), this method produces the following error:

```
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/mail/models/mail_mail.py", line 278, in send
    res = IrMailServer.send_email(msg, mail_server_id=mail.mail_server_id.id)
  File "/opt/odoo/auto/addons/mail_tracking/models/ir_mail_server.py", line 88, in send_email
    tracking_email.smtp_error(self, smtp_server_used, e)
  File "/opt/odoo/auto/addons/mail_tracking/models/mail_tracking_email.py", line 213, in smtp_error
    self.sudo()._partners_email_bounced_set('error')
  File "/opt/odoo/auto/addons/mail_tracking/models/mail_tracking_email.py", line 203, in
_partners_email_bounced_set
    ]).email_bounced_set(self, reason, event=event)
  File "/opt/odoo/auto/addons/mail_tracking_mailgun/models/res_partner.py", line 22, in
email_bounced_set
    self._email_bounced_set(reason, event)
  File "/opt/odoo/auto/addons/mail_tracking_mailgun/models/res_partner.py", line 33, in
_email_bounced_set
    event['Message-Id'] or '') TypeError: 'NoneType' object has no attribute '__getitem__'
```

So, we now assume we do not always have an event.

yajo commented:
This should be ready now. I took the freedom to amend @chienandalu's commit as per review.

rafaelbn reviewed (approved):
👍 

Please @Yajo @pedrobaeza review

pedrobaeza commented:
@Yajo please apply the same to v9
